x/shield/keeper/purchase.go

```diff
@@ -77,54 +77,61 @@ func (k Keeper) DequeuePurchase(ctx sdk.Context, purchaseList types.PurchaseList
 	pool, found := k.GetPool(ctx, poolID)
 	if !found {
 		return types.Purchase{}, types.ErrNoPoolFound
 	}
-	poolParams := k.GetPoolParams(ctx)
-	claimParams := k.GetClaimProposalParams(ctx)
-
-	// check preconditions
-	if err := k.PoolShieldAvailable(ctx, pool); err != nil {
-		return types.Purchase{}, err
+	if !pool.Active {
+		return types.Purchase{}, types.ErrPoolInactive
 	}
 
+	// Check available collaterals.
 	shieldAmt := shield.AmountOf(k.sk.BondDenom(ctx))
-	if shieldAmt.GT(pool.Available) {
-		return types.Purchase{}, types.ErrNotEnoughShield
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalWithdrawing := k.GetTotalWithdrawing(ctx)
+	totalShield := k.GetTotalShield(ctx)
+	if totalShield.Add(shieldAmt).GT(totalCollateral.Sub(totalWithdrawing)) {
+		return types.Purchase{}, types.ErrNotEnoughCollateral
 	}
-	if pool.SponsorAddr.Equals(purchaser) {
-		return types.Purchase{}, types.ErrSponsorPurchase
+
+	// Check pool shield limit.
+	poolParams := k.GetPoolParams(ctx)
+	maxShield := totalCollateral.Sub(totalWithdrawing).ToDec().Mul(poolParams.PoolShieldLimit).TruncateInt()
+	if shieldAmt.GT(maxShield) {
+		return types.Purchase{}, types.ErrPoolShieldExceedsLimit
 	}
 
-	// send tokens to shield module account
-	shieldDec := sdk.NewDecCoinsFromCoins(shield...)
-	premium, _ := shieldDec.MulDec(poolParams.ShieldFeesRate).TruncateDecimal()
-	if err := k.DepositNativePremium(ctx, premium, purchaser); err != nil {
+	// Send service fees to the shield module account.
+	if err := k.DepositNativeServiceFees(ctx, serviceFees, purchaser); err != nil {
 		return types.Purchase{}, err
 	}
 
-	// update pool premium, shield and available
-	premiumMixedDec := types.NewMixedDecCoins(sdk.NewDecCoinsFromCoins(premium...), sdk.DecCoins{})
-	pool.Premium = pool.Premium.Add(premiumMixedDec)
-	pool.Shield = pool.Shield.Add(shield...)
-	pool.Available = pool.Available.Sub(shieldAmt)
+	// Update global pool and project pool's shield.
+	totalShield = totalShield.Add(shieldAmt)
+	pool.Shield = pool.Shield.Add(shieldAmt)
+	k.SetTotalShield(ctx, totalShield)
 	k.SetPool(ctx, pool)
 
-	// set purchase
+	// Set a new purchase.
 	protectionEndTime := ctx.BlockTime().Add(poolParams.ProtectionPeriod)
-	claimPeriodEndTime := ctx.BlockTime().Add(claimParams.ClaimPeriod)
 	purchaseID := k.GetNextPurchaseID(ctx)
-	purchase := types.NewPurchase(purchaseID, shield, ctx.BlockHeight(), protectionEndTime, claimPeriodEndTime, claimPeriodEndTime, description)
+	purchase := types.NewPurchase(purchaseID, protectionEndTime, description, shieldAmt)
 	purchaseList := types.NewPurchaseList(poolID, purchaser, []types.Purchase{purchase})
 	k.AddPurchase(ctx, poolID, purchaser, purchase)
-	k.InsertPurchaseQueue(ctx, purchaseList, claimPeriodEndTime)
+	k.InsertPurchaseQueue(ctx, purchaseList, protectionEndTime.Add(k.GetPurchaseDeletionPeriod(ctx)))
```

Comment:
I don't see the point of using type `PurchaseList` here, but we can deal with this later